### PR TITLE
NCS-87 company shareholder count validation

### PR DESCRIPTION
**Pull request by AlfredBwomezi**
Add company shareholder count validation.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/client/OracleQueryClient.java [hunk 1]
```diff
@@ -29,4 +29,15 @@ public Long getCompanyTradedStatus(String companyNumber) {
...
+
+    public int getShareholderCount(String companyNumber) {
+        var shareholderCountUrl = String.format("%s/company/%s/shareholders/count", oracleQueryApiUrl, companyNumber);
+        LOGGER.info("Calling Oracle Query API URL (get): {}", shareholderCountUrl);
+
+        ResponseEntity<Integer> response = restTemplate.getForEntity(shareholderCountUrl, Integer.class);
+        var count = response.getBody();
+        LOGGER.info("Received {} from Oracle Query API URL (get): {}", count, shareholderCountUrl);
+
+        return count;
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/client/OracleQueryClient.java:41]: you are returning count here which is of type Integer, but the return type for this method is int.

If somehow the response body was null, we would get a null pointer exception here.
I would suggest either setting the return type to Integer. Or add a null check within this method

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyShareholderCountValidation.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    private final ShareholderService shareholderService;
+
+    public CompanyShareholderCountValidation(ShareholderService shareholderService) {
+        this.shareholderService = shareholderService;
+    }
+
+    @Override
+    public void validate(CompanyProfileApi companyProfile) throws EligibilityException {
+
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyShareholderCountValidation.java:23]: would you be able to add a feature flag similar to the Psc validation rule, that defaults to true. To allow us to bypass this eligibility checkif need be?

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ShareholderService.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.confirmationstatementapi.client.OracleQueryClient;
+
+@Service
+public class ShareholderService {
+
+    @Autowired
+    private OracleQueryClient oracleQueryClient;
+
+    public int getShareholderCount(String companyNumber) {
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ShareholderService.java:13]: the Integer returned from getShareholderCount is being unboxed here into an int.
I would suggest returning an Integer from here and add a null check to count in the company shareholder validation

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/configuration/ConfirmationStatementServiceEligibilityConfig.java [hunk 4]
```diff
@@ -45,25 +47,29 @@
...
         var companyOfficerValidation = new CompanyOfficerValidation(officerService, officerValidationFlag);
         var companyPscCountValidation = new CompanyPscCountValidation(pscService, pscValidationFeatureFlag);
         var companyTradedStatusValidation = new CompanyTradedStatusValidation(corporateBodyService);
+        var companyShareholderValidation = new CompanyShareholderCountValidation(shareholderService);
 
         listOfRules.add(companyStatusValidation);
         listOfRules.add(companyTypeValidationNoCS01Required);
         listOfRules.add(companyTypeValidationForWebFiling);
         listOfRules.add(companyTypeValidationPaperOnly);
+        listOfRules.add(companyTradedStatusValidation);
```
[the remark this answers is not in the document]
Comment (AlfredBwomezi) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/configuration/ConfirmationStatementServiceEligibilityConfig.java:67]: Is that (very last) after everything including the Officer checks?

[the remark this answers is not in the document]
Comment (ICoxCH) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/configuration/ConfirmationStatementServiceEligibilityConfig.java:67]: I'm basing that off of NCS-87, at the bottom of the story: "Check 4 (trading status of shares) will take place after check 3 is completed"

The PSC, Officers and Shareholders are all Check 3

**Review by ICoxCH — APPROVED** [review]

**Pull request merged**